Src/Support/GoogleApis.Tests/Apis/Requests/PageStreamerTest.cs

```diff
@@ -0,0 +1,227 @@
+            Assert.AreEqual(resource.AllItems, actual);
+        }
+
+        // Note: this doesn't test all the possible timings of cancellation
+        [Test]
+        public async Task Cancellation()
+        {
+            var cts = new CancellationTokenSource();
+            var gatekeeperSource = new TaskCompletionSource<int>();
+            var resource = new PageStreamedResource(
+                "check",
+                new Page("x", 1, 2, 3),
+                new Page("y", 4, 5),
+                new Page(null, 6, 7));
+            resource.gateKeeper = gatekeeperSource.Task;            
+
+            var task = PageStreamedResource.PageStreamer.FetchAllAsync(new Request(resource), cts.Token);
+            // Delay the resource fetching until we've cancelled the token
+            cts.Cancel();
+            gatekeeperSource.SetResult(0);
+            // TODO: Move to Assert.ThrowsAsync when we update to an appropriate version of NUnit.
+            // Assert.That is available, but the documentation is unclear how this interacts with await.
+            try
+            {
+                await task;
+                Assert.Fail("Expected exception");
+            }
+            catch (OperationCanceledException)
+            {
+                // Expected
+            }
+        }
+
+        // Has to be public so we can use it as a parameter for test cases
+        public class PageStreamedResource
+        {
+            internal static readonly PageStreamer<int, Request, Page, string> PageStreamer =
+                new PageStreamer<int, Request, Page, string>(
+                    (request, token) => request.Token = token,
+                    page => page.NextPageToken,
+                    page => page.Items);
+
+            // Task to await when fetching pages; defaults to an already completed task,
+            // but can be set to another task to allow for more fine-grained interaction.
+            internal Task gateKeeper = TaskEx.FromResult(0);
+            internal List<Page> Pages { get; private set; }
+            internal string Name { get; set; }
+
+            internal PageStreamedResource(string name, params Page[] pages)
+            {                
+                this.Name = name;
+                this.Pages = pages.ToList();
+                for (int i = 0; i < pages.Length - 1; i++)
+                {
+                    Pages[i].NextPageToken = Pages[i + 1].CurrentPageToken;
+                }
+            }
+
+            public override string ToString()
+            {
+                return Name;
+            }
+
+            internal Page GetPage(string token)
+            {
+                // This will fail if we provide a token that doesn't exist.
+                return Pages.Single(p => p.CurrentPageToken == token);
+            }
+
+            internal async Task<Page> GetPageAsync(string token, CancellationToken cancellationToken)
+            {
+                await gateKeeper;
```

Comment:
I just want to make sure we actually wait for this task... as mentioned above, it's a completed task, doesn't there is an optimization in TPL, that we will just continue in the asme thread, and then the whole multithread test is worth less?